### PR TITLE
feat(ui): the toolbox became vertical

### LINKED ISSUE #1972: [Bug] hover toolbar doesn't work on mobile responsive size layout
Describe a bug.
When interacting with the editor on my laptop chrome browser at the mobile responsive breakpoint, I can not hover the toolbar as the toolbar jumps to the next block when I attempt to hover. 

Steps to reproduce:
2. Go to editorjs.io homepage (I'm seeing this bug in my project as well as the official editor JS demo)
3. Reduce browser small enough for the mobile responsive layout to be active. 
4. Attempt to hover over the tool bar
5. The toolbar jumps to block below current block

Workaround: occasionally if you hover on an off enough times, it will allow you to hover. Also, if you try to select part of the text in an element, you can get it to hover (see last part of screenshot)

Expected behavior:
I can hover over the toolbar for the current element. 

Screenshots:
Feb-16-2022 23-13-04

Device, Browser, OS:
Macbook Pro, Mac OS Monterey, Chrome Version 98.0.4758.80 

Editor.js version:
Current version live on editorjs.io

### PR DESCRIPTION
**Pull request by neSpecc**
This PR is the second part of the big UI update called "Unified Toolbox". See the [Roadmap](https://github.com/codex-team/editor.js#roadmap) for more details.

In this iteration, the Toolbox became a vertical list.

image

image

## Changelog

- the `Popover` utility has been created. It will be used for all vertical lists.
- the `Toolbox` now uses the `Popover` to display the tools list
- the `Plus` button will be always shown (previously, it appears only for empty blocks) 

image

- the `Toolbar` won't move on hover at mobile viewports. Resolves #1972 
- the `Flipper` now scrolls to the focused item using `scrollIntoViewIfNeeded`

---

Also, see the Discussion https://github.com/codex-team/editor.js/discussions/1811

---

Thanks to @TatianaFomina for finishing the PR so fast.

### CODE REVIEW

File: src/components/flipper.ts [hunk 1]
```diff
@@ -266,4 +266,15 @@ export default class Flipper {
...
     event.stopPropagation();
   }
+
+  /**
+   * Fired after flipping in any direction
+   */
+  private flipCallback(): void {
+    if (this.iterator.currentItem) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+      // @ts-ignore
```
Comment (gohabereg) [inline src/components/flipper.ts]: Why ts-ignore here?

File: src/components/modules/toolbar/index.ts [hunk 2]
```diff
@@ -490,6 +509,15 @@ export default class Toolbar extends Module<ToolbarNodes> {
...
      */
     this.eventsDispatcher.on(this.Editor.UI.events.blockHovered, (data: {block: Block}) => {
+      /**
+       * Do not move Toolbar by hover on mobile view
+       *
+       * @see https://github.com/codex-team/editor.js/issues/1972
+       */
+      if (_.isMobile()) {
+        return;
+      }
```
Comment (gohabereg) [inline src/components/modules/toolbar/index.ts]: Can we just do not subscribe on mobile?

File: src/components/polyfills.ts [hunk 3]
```diff
@@ -96,3 +96,40 @@ if (!Element.prototype.prepend) {
...
+ *
+ * @see {@link https://gist.github.com/KilianSSL/774297b76378566588f02538631c3137}
+ */
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore
+if (!Element.prototype.scrollIntoViewIfNeeded) {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
+  Element.prototype.scrollIntoViewIfNeeded = function (centerIfNeeded): void {
+    centerIfNeeded = arguments.length === 0 ? true : !!centerIfNeeded;
```
Comment (gohabereg) [inline src/components/polyfills.ts]: You can extends `Element` interface to avoid ts-ignore here

File: src/components/ui/toolbox.ts [hunk 4]
```diff
@@ -124,67 +115,79 @@ export default class Toolbox extends EventsDispatcher<ToolboxEvent> {
...
+            this.toolButtonActivated(tool.name);
+          },
+          secondaryLabel: tool.shortcut ? _.beautifyShortcut(tool.shortcut) : '',
+        };
+      }),
+    });
 
-    this.addTools();
-    this.enableFlipper();
+    this.popover.on(PopoverEvent.OverlayClicked, () => {
```
Comment (gohabereg) [inline src/components/ui/toolbox.ts]: Do we unsubscribe somewhere?

File: src/components/utils.ts [hunk 5]
```diff
@@ -762,3 +762,10 @@ export function cacheable<Target, Value, Arguments extends unknown[] = unknown[]
 
   return descriptor;
 };
+
+/**
+ * True if screen has mobile size
+ */
+export function isMobile(): boolean {
```
Comment (gohabereg) [inline src/components/utils.ts]: I suggest to rename the function as it doesn't return what expected. Something like `isNarrowScreen`

File: src/components/utils/popover.ts [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+      itemHidden: 'ce-popover__item--hidden',
+      itemFocused: 'ce-popover__item--focused',
+      itemLabel: 'ce-popover__item-label',
+      itemIcon: 'ce-popover__item-icon',
+      itemSecondaryLabel: 'ce-popover__item-secondary-label',
+      noFoundMessage: 'ce-popover__no-found',
+      noFoundMessageShown: 'ce-popover__no-found--shown',
+      popoverOverlay: 'ce-popover__overlay',
+      popoverOverlayHidden: 'ce-popover__overlay--hidden',
+      documentScrollLocked: 'ce-scroll-locked',
```
Comment (gohabereg) [inline src/components/utils/popover.ts:145]: We should really think of using css modules

File: src/components/utils/popover.ts [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+  /**
+   * Renders the single item
+   *
+   * @param item - item data to be rendered
+   */
+  private createItem(item: PopoverItem): HTMLElement {
+    const el = Dom.make('div', Popover.CSS.item);
+
+    el.setAttribute('data-item-name', item.name);
```
Comment (gohabereg) [inline src/components/utils/popover.ts]:
```suggestion
    el.dataset.itemName =  item.name;
```

File: src/components/utils/search-input.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  /**
+   * Creates the search field
+   *
+   * @param placeholder - input placeholder
+   */
+  private render(placeholder: string): void {
+    this.wrapper = Dom.make('div', SearchInput.CSS.wrapper);
+    const iconWrapper = Dom.make('div', SearchInput.CSS.icon);
```
Comment (gohabereg) [inline src/components/utils/search-input.ts:116]:
```suggestion

    const iconWrapper = Dom.make('div', SearchInput.CSS.icon);
```

File: src/styles/input.css [hunk 9]
```diff
@@ -0,0 +1,46 @@
+img, video, audio, canvas, input,
```
Comment (gohabereg) [inline src/styles/input.css]: Not encapsulated selectors

**Review by gohabereg — CHANGES_REQUESTED** [review]

File: src/components/modules/toolbar/index.ts (hunk 2 above)
Comment (TatianaFomina) [reply src/components/modules/toolbar/index.ts]: fixed

File: src/components/polyfills.ts (hunk 3 above)
Comment (TatianaFomina) [reply src/components/polyfills.ts]: fixed

File: src/components/ui/toolbox.ts (hunk 4 above)
Comment (TatianaFomina) [reply src/components/ui/toolbox.ts]: fixed

File: src/components/utils.ts (hunk 5 above)
Comment (TatianaFomina) [reply src/components/utils.ts]: renamed to `isMobileScreen`

File: src/components/flipper.ts [hunk 10]
```diff
@@ -206,23 +206,7 @@ export default class Flipper {
...
-      handlingKeyCodeList.push(
-        _.keyCodes.LEFT,
-        _.keyCodes.RIGHT,
-        _.keyCodes.UP,
-        _.keyCodes.DOWN
-      );
-    }
-
-    return this.activated && handlingKeyCodeList.indexOf(event.keyCode) !== -1;
+    return this.activated && this.allowedKeys.indexOf(event.keyCode) !== -1;
```
Comment (gohabereg) [inline src/components/flipper.ts]:
```suggestion
    return this.activated && this.allowedKeys.includes(event.keyCode);
```

File: src/components/utils/popover.ts [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+      el.appendChild(Dom.make('div', Popover.CSS.itemIcon, {
+        innerHTML: item.icon,
+      }));
+    }
+
+    el.appendChild(label);
+
+    if (item.secondaryLabel) {
+      el.appendChild(Dom.make('div', Popover.CSS.itemSecondaryLabel, {
+        innerHTML: item.secondaryLabel,
```
Comment (gohabereg) [inline src/components/utils/popover.ts]: Do we really need `innerHTML` here? Can we use `textCintent`?

File: src/styles/input.css (hunk 9 above)
Comment (gohabereg) [reply src/styles/input.css]: Still relevant

Comment (TatianaFomina) [reply src/styles/input.css]: removed these lines

File: src/components/utils/popover.ts (hunk 6 above)
Comment (neSpecc) [reply src/components/utils/popover.ts:145]: We will, at some of the next updates

**Review by gohabereg — APPROVED** [review]

**Review by ilyamore88 — APPROVED** [review]

**Pull request merged**